### PR TITLE
NATS pubsub component implementation

### PR DESCRIPTION
**Pull request by abhirockzz**
# Description

Add support for [NATS](https://github.com/nats-io/nats-server) as a pubsub component in Dapr. To be specific, the intention is to use [NATS queue groups](https://docs.nats.io/nats-concepts/queue) such that consumer Dapr apps can share the workload in a load-balanced way.

## Issue reference

Issue - #64 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by yaron2** [comment]
Hey,

To test an integration with the Dapr runtime, first reference your local version of components-contrib using `go.mod` in dapr/dapr.

Add the following line at the bottom:
`replace github.com/dapr/components-contrib => ../components-contrib`

Then run: `go mod vendor`.

This should pull all the dependencies including components-contrib from your local machine.

Then, update the loader in the Dapr runtime under `/pkg/components/pubsub/loader.go`:

```
RegisterMessageBus("nats", nats. NewNATSPubSub)
```

You can then build the daprd binary locally and test it with the [pub-sub sample
](https://github.com/dapr/samples/tree/master/4.pub-sub).

File: pubsub/nats/nats.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		return fmt.Errorf("nats: error connecting to nats at %s: %s", m.natsURL, err)
+	}
+
+	n.natsConn = natsConn
+	return nil
+}
+
+func (n *natsPubSub) Publish(req *pubsub.PublishRequest) error {
+	fmt.Printf("nats Publish request on subject %s with data %s", req.Topic, string(req.Data))
```
Comment (yaron2) [inline pubsub/nats/nats.go]: Can you make this `log.Debugf`?

File: pubsub/nats/nats.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	err := n.natsConn.Publish(req.Topic, req.Data)
+	if err != nil {
+		return fmt.Errorf("nats: error from publish: %s", err)
+	}
+	fmt.Println("message published successfully to NATS")
```
Comment (yaron2) [inline pubsub/nats/nats.go]: Can you make this `log.Debugf`?

File: pubsub/nats/nats.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+	return nil
+}
+
+func (n *natsPubSub) Subscribe(req pubsub.SubscribeRequest, handler func(msg *pubsub.NewMessage) error) error {
+	fmt.Printf("nats Subscribe request on subject %s", req.Topic)
```
Comment (yaron2) [inline pubsub/nats/nats.go]: Can you make this `log.Debugf`?

File: pubsub/nats/nats.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	sub, err := n.natsConn.QueueSubscribe(req.Topic, n.metadata.queue, func(natsMsg *nats.Msg) {
+		handler(&pubsub.NewMessage{Topic: req.Topic, Data: natsMsg.Data})
+	})
+	if err != nil {
+		log.Warnf("nats: error subscribe: %s", err)
+	}
+	fmt.Printf("NATS Subscribe request successful %v", sub)
```
Comment (yaron2) [inline pubsub/nats/nats.go]: Can you make this `log.Debugf`?

**Comment by abhirockzz** [comment]
Thanks @yaron2 . I had already tried something similar to what you mentioned but ran into issues. Re-tried again with what you outlined and facing the same problems. Here is what I did

- `git clone https://github.com/dapr/dapr`
- build worked - `cd dapr && make build`
- modified `go.mod` - added `replace github.com/dapr/components-contrib => ../components-contrib` at the end
- `go mod vendor` (no code change to include NATS pubusb yet)
- `make build` failed

Build error:

```
build command-line-arguments: cannot load github.com/Azure/go-autorest/autorest: open /Users/abhishekgupta/go/src/github.com/dapr/dapr/vendor/github.com/Azure/go-autorest/autorest: no such file or directory
make: *** [dist/darwin_amd64/release/daprd] Error 1
```

**Comment by abhirockzz** [comment]
I managed to test this end-to-end with Dapr runtime "with hacks" as of now. [Here is an example](https://github.com/abhirockzz/dapr-nats-pubsub)

**Comment by youngbupark** [comment]
@abhirockzz please fix merge conflict.

**Comment by yaron2** [comment]
I'm excited to see this merged!

**Comment by yaron2** [comment]
Tests failing 😣

**Comment by yaron2** [comment]
Can you run `go mod vendor` and push the vendored files?

**Comment by yaron2** [comment]
Seems like there's still some missing vendor deps.

Can you run make lint locally and make sure it passes?

**Comment by abhirockzz** [comment]
I tried `make test` and `make lint`

result for `make lint`

```
# Due to https://github.com/golangci/golangci-lint/issues/580, we need to add --fix for windows
golangci-lint run --fix
ERRO Running error: context loading failed: package github.com/Azure/go-autorest/autorest/azure: no go files to analyze
make: *** [lint] Error 5
```

result for `make test`

```
go test ./... -mod=vendor
vendor/github.com/Azure/azure-sdk-for-go/services/eventhub/mgmt/2017-04-01/eventhub/client.go:24:2: cannot find package "." in:
	/Users/abhishekgupta/go/src/github.com/dapr/components-contrib/vendor/github.com/Azure/go-autorest/autorest
vendor/github.com/Azure/azure-amqp-common-go/aad/jwt.go:37:2: cannot find package "." in:
	/Users/abhishekgupta/go/src/github.com/dapr/components-contrib/vendor/github.com/Azure/go-autorest/autorest/adal
vendor/github.com/Azure/azure-amqp-common-go/aad/jwt.go:38:2: cannot find package "." in:
	/Users/abhishekgupta/go/src/github.com/dapr/components-contrib/vendor/github.com/Azure/go-autorest/autorest/azure
vendor/github.com/Azure/azure-event-hubs-go/atom/atom.go:7:2: cannot find package "." in:
	/Users/abhishekgupta/go/src/github.com/dapr/components-contrib/vendor/github.com/Azure/go-autorest/autorest/date
vendor/github.com/Azure/azure-sdk-for-go/services/eventhub/mgmt/2017-04-01/eventhub/models.go:26:2: cannot find package "." in:
	/Users/abhishekgupta/go/src/github.com/dapr/components-contrib/vendor/github.com/Azure/go-autorest/autorest/to
vendor/github.com/Azure/azure-sdk-for-go/services/eventhub/mgmt/2017-04-01/eventhub/consumergroups.go:24:2: cannot find package "." in:
	/Users/abhishekgupta/go/src/github.com/dapr/components-contrib/vendor/github.com/Azure/go-autorest/autorest/validation
vendor/github.com/Azure/azure-sdk-for-go/services/eventhub/mgmt/2017-04-01/eventhub/consumergroups.go:25:2: cannot find package "." in:
	/Users/abhishekgupta/go/src/github.com/dapr/components-contrib/vendor/github.com/Azure/go-autorest/tracing
secretstores/keyvault/authutils.go:16:2: cannot find package "." in:
	/Users/abhishekgupta/go/src/github.com/dapr/components-contrib/vendor/github.com/Azure/go-autorest/autorest/azure/auth
make: *** [test] Error 1
```

**Comment by yaron2** [comment]
Ok I see the issue with your `go.mod`.

Add the following:
```
github.com/Azure/go-autorest v13.0.1+incompatible // indirect
```

And remove `// indirect` from:
```
github.com/Azure/go-autorest/autorest v0.9.1
```

**Comment by yaron2** [comment]
LGTM

**Comment by abhirockzz** [comment]
> Ok I see the issue with your `go.mod`.
> 
> Add the following:
> 
> ```
> github.com/Azure/go-autorest v13.0.1+incompatible // indirect
> ```
> 
> And remove `// indirect` from:
> 
> ```
> github.com/Azure/go-autorest/autorest v0.9.1
> ```
> 
> Run `go mod vendor` again.

Thanks for the help!

**Comment by yaron2** [comment]
Sure thing

**Comment by liangyuanpeng** [comment]
> Hey,
> 
> To test an integration with the Dapr runtime, first reference your local version of components-contrib using `go.mod` in dapr/dapr.
> 
> Add the following line at the bottom:
> `replace github.com/dapr/components-contrib => ../components-contrib`
> 
> Then run: `go mod vendor`.
> 
> This should pull all the dependencies including components-contrib from your local machine.
> 
> Then, update the loader in the Dapr runtime under `/pkg/components/pubsub/loader.go`:
> 
> ```
> RegisterMessageBus("nats", nats.NewNATSPubSub)
> ```
> 
> You can then build the daprd binary locally, copy it to your PATH and test it with the [pub-sub sample ](https://github.com/dapr/samples/tree/master/4.pub-sub)

I noticed the need to change load.go，So is there no CLI or openapi to change this now?

or Maybe have some roadmap for this

**Comment by abhirockzz** [comment]
@liangyuanpeng This was recently incorporated by @yaron2 (see [this PR](https://github.com/dapr/dapr/pull/716)) - you should see it now. This is part of the [November 2019 milestone](https://github.com/dapr/dapr/milestone/5)

**Comment by derekcollison** [comment]
This is great work! I am excited to see this. Quick question, is the logging dependency still supposed to be there?

Also, as Dapr newbie, does the pub/sub system support wildcards? In NATS you can subscribe to wildcard subjects, like `foo.*` which matches `foo.1`, `foo.2` etc. and also `foo.>` which matches any subject, no matter how many tokens, that starts with `foo.`.

Reason I bring it up is the impl hands back the subscribed topic and not the one attached to the message, which when using wildcards would be different.

Would also love to see support for credentials, so maybe I will try to add that with a quick PR.

Thanks again!

**Comment by yaron2** [comment]
hey @derekcollison, from what I can see in the implementation of NATS, the answer is yes, wildcards are supported. There's nothing in Dapr itself that prevents any format of topics from being used.

It'd be great if you could add support for credentials!

**Comment by yaron2** [comment]
Updated answer

**Comment by derekcollison** [comment]
ok sounds good. Will wait on @ abhirockzz for an answer on the log debug print etc.

Thanks..

**Comment by abhirockzz** [comment]
@derekcollison thanks for your inputs!

Regarding wildcard subscriptions

Technically speaking, the subscriber will work as is e.g. if you setup everything (point Dapr to `demo.nats.io:4222`), subscribe to e.g. (`foo.*`) and then publish a message to topic `foo.bar`, you will see that the application will get the message (via Dapr).

The above is correct for a Dapr "binding", but, pubsub is for inter-app communication. So, the publishing needs to happen through another Dapr app (rather than directly via `nats`), and that's where the catch is. The publish API simply uses the topic "name" (in Dapr) i.e. `foo.*` in this case.
(@yaron2 please correct me if I am missing anything?)

> [This example (wip)](https://github.com/abhirockzz/dapr-nats-pubsub) should give you an idea.

Regarding logging, are you referring to the `log.Debugf` statements (using `logrus`)?

**Comment by derekcollison** [comment]
In publish/subscribe systems like NATS you can subscribe to a wildcard but you can only publish to literals. Meaning I can subscribe to `foo.*` and publish to `foo.22` or `foo.33` and the subscriber will get both.

In regards to logging, yes was interested in why you would include a logging library in a client API.

**Comment by abhirockzz** [comment]
@derekcollison Wildcard subjects work as well. I tried it and updated [the instructions here](https://github.com/abhirockzz/dapr-nats-pubsub#wildcard-subscriptions).

**Comment by derekcollison** [comment]
Great! Thanks.

**Pull request merged**